presto-main/src/main/java/com/facebook/presto/execution/SetSessionTask.java

```diff
@@ -35,8 +46,33 @@ public void execute(SetSession statement, Session session, Metadata metadata, Qu
         if (statement.getName().getParts().size() > 2) {
             throw new SemanticException(INVALID_SESSION_PROPERTY, statement, "Invalid session property '%s'", statement.getName());
         }
+        String name = statement.getName().toString();
 
-        // todo verify session properties are valid
-        stateMachine.addSetSessionProperties(statement.getName().toString(), statement.getValue());
+        SessionPropertyMetadata<?> sessionPropertyInfo = metadata.getSessionPropertyManager().getSessionPropertyMetadata(name);
+        Type type = sessionPropertyInfo.getSqlType();
+
+        Object objectValue = toObjectValue(statement.getValue(), type, session, metadata);
+        String value = serializeSessionProperty(type, objectValue);
+
+        // verify the sql value can be decoded by the property
+        metadata.getSessionPropertyManager().decodeProperty(name, value, sessionPropertyInfo.getJavaType());
+
+        stateMachine.addSetSessionProperties(name, value);
+    }
+
+    @NotNull
+    private static Object toObjectValue(Expression expression, Type expectedType, Session session, Metadata metadata)
+    {
+        Object value = evaluateConstantExpression(expression, expectedType, metadata, session);
+
+        // convert to object value type of SQL type
+        BlockBuilder blockBuilder = expectedType.createBlockBuilder(new BlockBuilderStatus(), 1);
+        BlockUtils.appendObject(expectedType, blockBuilder, value);
+        Object objectValue = expectedType.getObjectValue(session.toConnectorSession(), blockBuilder, 0);
+
+        if (objectValue == null) {
+            throw new PrestoException(StandardErrorCode.INVALID_SESSION_PROPERTY, "Session property value can not be null");
```

Comment:
It can not be static imported because `com.facebook.presto.sql.analyzer.SemanticErrorCode.INVALID_SESSION_PROPERTY` is already imported for another exception